### DISCUSSION TITLE
[V5] Dynamic connection depending on route parameter?

### DISCUSSION THREAD

**Discussion by mikelogaciuk** [adonisjs/core#2332] [Help] [answered]
I am trying to build up connection depending on route parameter and tried to do so on my own but it didn't go well.

I have few servers with the same schema but different hostnames. I would like to update a connection string from database.ts depending on endpoint parameter,

I've tried something like this but it didn't worked:

```ts

Route.get('/stores/:store', async ({ params }) => {
  Env.set('STORE_HOST', `${params.store}.domain.subfix`)

  Logger.info(Env.get('STORE_HOST'))

  const market = await Database.connection('stores').rawQuery(
    'SELECT * FROM sales FETCH NEXT 1000 ROWS ONLY'
  )

  return { market }
})
```

I've checked and every REQ changes 'STORE_HOST' but still knex is returning me an err:

`Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a 
.transacting(trx) call?`

Is there any way to pass `{params}` into `connection: {host}`? Or should I use separate route file with separate connection / database builder?

Any snippet would be nice.

**Comment by McSneaky** [comment]
Hmm, yea, changing Env variables at runtime wont end well
Can you predefine connections on Database.ts and then use them based on query param?

Then you could do something like this:
```js
await Database.connection(`${params.store}.stores`)....
```

**Comment by vookio** [comment]
What you are looking for is in "[on demand connections](https://preview.adonisjs.com/guides/database/connections-management#on-demand-connections)" section I think. 

You can use the mentioned middleware to register the connections once, and [switch between the connections](https://preview.adonisjs.com/guides/database/connections-management#switching-between-connections) in the routes.

**Reply by mikelogaciuk** [reply]
I figured that out on my own before I read your tip. Wasn't looking and thinking. 

Thanks for your initiative.

**Comment by mikelogaciuk** [comment, marked as answer]
Middleware and controller didn't work for me as I wanted, but I figured it differently. I used direct code in route and it worked like a charm. 

Mock code can be found below:

```ts
import Env from '@ioc:Adonis/Core/Env'
import Route from '@ioc:Adonis/Core/Route'
import Database from '@ioc:Adonis/Lucid/Database'
import Logger from '@ioc:Adonis/Core/Logger'

Route.get('/stores/stock/:store/:product', async ({ params }) => {
  const storeConnection = `${params.store}ENTITY`
  const product = `${params.product}`

  Logger.info(`Product stock in store ${storeConnection} requested for product: ${product}`)

  const storeConnectionConfig = {
    client: 'oracledb' as const,
    connection: {
      host: storeConnection,
      port: Env.get('STORE_PORT'),
      user: Env.get('STORE_USER'),
      password: Env.get('STORE_PASSWORD'),
      database: Env.get('STORE_DB_NAME'),
    },
  }

  if (!Database.manager.has(storeConnection)) {
    Database.manager.add(storeConnection, storeConnectionConfig)
    Database.manager.connect(storeConnection)
  }

  const result = await Database.connection(storeConnection).rawQuery(
    `SELECT
          product
          ,store
          ,quantity
      FROM stocks
      WHERE product = ?`,
    [product]
  )
  return result

  await Database.manager.close(storeConnection, true)
})
```
